### PR TITLE
[Feat] Add New Postproc (Cut Road Connection)

### PR DESCRIPTION
**Pull request by geoyee**
### PR types

New features

### PR changes

Others 

### Description

初步添加一个道路断线连接的后处理方法。参考论文(兼顾连通性的U-Net网络高分辨率遥感影像道路提取)[http://hgs.publish.founderss.cn/thesisDetails?columnId=4759509&lang=zh]，这个也没公开的代码，实现可能与原文有出入。

论文提取的道路结果：

test

论文连接断线的结果：

{7EPCU I_AQ(R(6OXQ5`))E

复现的结果：

4Y6(~E2U$(26N_~H8{V6% 4

右侧有个错误的连接，我看是kmeans得到的结果不同，不知道他是怎样选择最优的类别数的。我使用的是平均轮廓系数最大。

使用方法

``` python
from PIL import Image
import numpy as np
import cv2
from paddlers.utils.postprocs import cut_road_connection

mask = np.asarray(Image.open("test.png"))
res = cut_road_connection(mask, line_width=6)
cv2.imwrite("output.png", res)
```

最后稍微加了下主README中这部分

### CODE REVIEW

File: paddlers/utils/postprocs/connection.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+
+warnings.filterwarnings("ignore", category=DeprecationWarning)
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 这个是为了滤除什么警告呢？

File: paddlers/utils/postprocs/connection.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+import itertools
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: itertools是标准库，按照[PaddleRS的模块导入格式规范](https://github.com/PaddlePaddle/PaddleRS/blob/develop/docs/CONTRIBUTING.md#22-%E6%A8%A1%E5%9D%97%E5%AF%BC%E5%85%A5%E9%A1%BA%E5%BA%8F)不应该和numpy、skimage等第三方库放在一起。

File: paddlers/utils/postprocs/connection.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+import cv2
+import numpy as np
+from skimage import morphology
+from sklearn import metrics
+from sklearn.cluster import KMeans
+from scipy import ndimage, optimize
+from .utils import prepro_mask, calc_distance
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py:28]: 同样，.utils也不应该和第三方库的import语句放在一起。

File: paddlers/utils/postprocs/connection.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+
+def cut_road_connection(mask: np.ndarray, line_width: int=6) -> np.ndarray:
+    """
+    Connection of cut road lines.
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 用verb-noun形式更好，`Connecting cut road lines.`

File: paddlers/utils/postprocs/connection.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+    The original article refers to
+    Wang B, Chen Z, et al. "Road extraction of high-resolution satellite remote sensing images in U-Net network with consideration of connectivity."
+    (http://hgs.publish.founderss.cn/thesisDetails?columnId=4759509).
+
+    This algorithm has no public code.
+    The implementation procedure refers to original article,
+    and it is not fully consistent with the article.
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 是否需要说明一下，具体有哪些不一致？

File: paddlers/utils/postprocs/connection.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+    Args:
+        mask (np.ndarray): Mask of road.
+        line_width (int, optional): Width of the line used for patching.
+            . Default is 6.
+
+    Returns:
+        np.ndarray: Mask of road after connected to cut road lines.
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: `Mask of road after connecting cut road lines.`

File: paddlers/utils/postprocs/connection.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    for point, lab in zip(break_points, labels):
+        if lab in match_points.keys():
+            match_points[lab].append(point)
+        else:
+            match_points[lab] = [point]
+    return match_points
+
+
+def _draw_curve(mask, skeleton, match_points, line_width):
+    result = mask.copy() * 255
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 这里`copy()`的意图是？考虑到`*255`的操作本身已经会创建一个新的array

File: paddlers/utils/postprocs/connection.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+    for _, v in match_points.items():
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 如果只用到`items()`中的value可以直接遍历`values()`

**Review by Bobholamovic — COMMENTED** [review]
完成度很高~ 一些小建议

File: paddlers/utils/postprocs/connection.py (hunk 1 above)
Comment (geoyee) [reply paddlers/utils/postprocs/connection.py]: 应该是`sklearn`和`numpy`的版本问题，像这样

```
D:\ProgramData\Anaconda3\envs\paddlers\lib\site-packages\sklearn\xxxxx\xxx.py:xx: DeprecationWarning: `np.float` is a deprecated 
alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
```

File: paddlers/utils/postprocs/connection.py (hunk 5 above)
Comment (geoyee) [reply paddlers/utils/postprocs/connection.py]: 已添加

File: paddlers/utils/postprocs/connection.py (hunk 1 above)
Comment (Bobholamovic) [reply paddlers/utils/postprocs/connection.py]: 这样的话我不建议一刀切（因为paddlers自动import这个模块后，可能会把其它地方本来可以暴露出来的警告也给埋没了），可以针对特定的模块处理一下，[warnings标准库是支持这个功能的](https://docs.python.org/zh-cn/3/library/warnings.html#the-warnings-filter)。

File: paddlers/utils/postprocs/regularization.py [hunk 9]
```diff
@@ -208,35 +200,35 @@ def _fine(contour, W):
...
+                                                   cur_edge_p2)
                 final_points.append(point_move)
                 # update next
                 correct_points[next_idx][0] = point_move
-                correct_points[next_idx][1] = calc_project_in_line(
+                correct_points[next_idx][1] = _calc_project_in_line(
                     next_edge_p2, cur_edge_p1, cur_edge_p2)
             else:
-                # add line
+                # add _line
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/regularization.py]: 这个地方的`_line`为啥加下划线呀？

File: paddlers/utils/postprocs/connection.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    Connecting cut road lines.
+
+    The original article refers to
+    Wang B, Chen Z, et al. "Road extraction of high-resolution satellite remote sensing images in U-Net network with consideration of connectivity."
+    (http://hgs.publish.founderss.cn/thesisDetails?columnId=4759509).
+
+    This algorithm has no public code.
+    The implementation procedure refers to original article,
+    and it is not fully consistent with the article.
+    1. It is an implementation of uncertainty about `n_clusters` of KMeans in the original article.
```
Comment (Bobholamovic) [inline paddlers/utils/postprocs/connection.py]: 看一下是不是要表达下述的意思：
and it is not fully consistent with the article:
1. The way to determine the optimal number of clusters `k` used in k-means clustering is not described in the original article. In this implementation, we use the `k` that reports the highest silhouette score.
2. We unmark the breakpoints if the angle between the two road extensions is less than 90°.

File: paddlers/utils/postprocs/regularization.py (hunk 9 above)
Comment (geoyee) [reply paddlers/utils/postprocs/regularization.py]: 替换的时候忘了...

File: paddlers/utils/postprocs/connection.py (hunk 10 above)
Comment (geoyee) [reply paddlers/utils/postprocs/connection.py]: 英语老大难了

File: paddlers/utils/postprocs/connection.py (hunk 1 above)
Comment (geoyee) [reply paddlers/utils/postprocs/connection.py]: 看是使用这个上下文管理器可以吗晖哥

**Review by Bobholamovic — APPROVED** [review]
LGTM.
